File: algorithms/309. Best Time to Buy and Sell Stock with Cooldown/solution.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        # refer: https://discuss.leetcode.com/topic/30421/share-my-thinking-process
        pre_sell, sell, buy = 0, 0, float('-inf')
        for p in prices:
            pre_sell, sell, buy = sell, max(buy + p, sell), max(pre_sell - p, buy)
        return sell

File: algorithms/309. Best Time to Buy and Sell Stock with Cooldown/solution.cc
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int prebuy = 0, presell = 0, buy = INT_MIN, sell = 0;
        for ( int p : prices ) {
            prebuy = buy;
            buy = max(presell - p, buy);
            presell = sell;
            sell = max(prebuy + p, sell);
        }
        return sell;
    }
};
